src/module/apps/trait-selector.js

```diff
@@ -9,120 +9,93 @@ export class TraitSelectorSFRPG extends FormApplication {
-        current = current.split(/[\s,]/).filter(t => !!t);
-        return current.map(val => {
-            for (let [k, v] of Object.entries(choices)) {
-                if (val === v) return k;
+    async filterTraits(li) {
+        li.hide();
+
+        for (const trait of li) {
+            if (this.searchMatch(trait)) {
+                $(trait).show();
             }
-            return null;
-        }).filter(val => !!val);
+        }
     }
 
     /**
-     * Update the Actor object with new trait data processed from the form
+     * Check if a displayed item matches the search string
      *
-     * @param {Event} event The event that triggers the update
-     * @param {Object} formData The data from the form
+     * @param {String} trait The localized trait name
      * @private
      */
-    _updateObject(event, formData) {
-        let choices = [];
+    searchMatch(trait) {
+        const searchTerm = this.searchTerm;
 
-        if (this.attribute !== "system.traits.dr") {
-            for (let [k, v] of Object.entries(formData)) {
-                if ((k !== 'custom') && v) choices.push(k);
-            }
-        } else {
-            let resistances = Object.entries(formData).filter(e => e[0].startsWith("er"));
-            resistances = resistances.reduce((obj, entry) => {
-                let [type, i] = entry[0].split('.').slice(1);
+        if (searchTerm !== '') {
+            const strings = this.searchTerm.split(',');
 
-                if (!obj[type]) obj[type] = {};
-                obj[type][i] = entry[1];
+            for (const string of strings) {
+                const textToSearch = $(trait).find('input')[0].nextSibling.data.toLowerCase().trim();
+                if (textToSearch.indexOf(string.toLowerCase().trim()) === -1) {
+                    return false;
+                }
+            }
+        }
 
-                return obj;
-            }, {});
+        return true;
+    }
 
-            choices = Object.entries(resistances).filter(e => e[1][0])
-                .reduce((arr, resistance) => {
-                    arr.push({[resistance[0]]: resistance[1][1]});
+    /**
+     * Act on inputs to the html
+     *
+     * @param {Object} html
+     */
+    activateListeners(html) {
 
-                    return arr;
-                }, []);
-        }
+        // activating or deactivating filters
+        html.on('change keyup paste', 'input[name=textFilter]', ev => {
+            this.searchTerm = ev.target.value;
+            this.filterTraits(html.find('li'));
+        });
 
-        this.object.update({
-            [`${this.attribute}.value`]: choices,
-            [`${this.attribute}.custom`]: formData.custom
+        // re-render the template with updated data if a trait is selected/unselected
+        html.on('change', 'input[type=checkbox]', async ev => {
+            const formData = this._getSubmitData();
+            await this._updateObject(ev, formData);
+            this.render();
```

Comment:
We shouldn't be updating the actor as soon as checkboxes are ticked, we should wait till the form is submitted. I understand why you're doing this, so you can just re-render it, but another solution should be used. Best I can suggest is a bunch of event listeners shuffling things into place as needed,

Reply:
I've updated this now to shuffle traits around using event listeners instead of always submitting